Implement encrypted push and pull/bud/from using encrypted images

lumjjb opened the pull request:
<!--
Thanks for sending a pull request!

Please make sure you've read and understood our contributing guidelines
(https://github.com/containers/buildah/blob/master/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.
-->

#### What type of PR is this?

<!--
Please label this pull request according to what type of issue you are
addressing, especially if this is a release targeted pull request.

Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

/kind feature

#### What this PR does / why we need it:

This PR introduces the ability to encrypt an OCI container image when pushing, and also allows the use of encrypted OCI container images when building a Dockerfile or pulling an encrypted image from the registry.

This is the initial implementation idea. There are a few more things that I am working on right now before taking out of DRAFT, but would like some feedback on the general idea.

TODO in progress by me:
- Remove ufave/cli and some additional imports which are not necessary from upstream ocicrypt lirbary
- Add tests

#### How to verify it

```
# Create keypair
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ openssl genrsa -out mykey.pem 1024
Generating RSA private key, 1024 bit long modulus (2 primes)
...........+++++
.......+++++
e is 65537 (0x010001)
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ openssl rsa -in mykey.pem -pubout > mykey.pub
writing RSA key

# Encrypt an image to local registry
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah images
REPOSITORY   TAG   IMAGE ID   CREATED   SIZE
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah pull docker://docker.io/library/nginx:latest
Getting image source signatures
Copying blob ffadbd415ab7 done
Copying blob c499e6d256d6 done
Copying blob 74cda408e262 done
Copying config ed21b7a8ae done
Writing manifest to image destination
Storing signatures
ed21b7a8aee9cc677df6d7f38a641fa0e3c05f65592c592c9f28c42b3dd89291
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                    NAMES
f91d02361eae        registry:2.7.1      "/entrypoint.sh /etc…"   4 months ago        Up 2 days           0.0.0.0:5000->5000/tcp   registry
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah push --tls-verify=false --encryption-key jwe:./mykey.pub docker.io/library/nginx:latest docker://localhost:5000/test_enc_img
Getting image source signatures
Copying blob d37eecb5b769 done
Copying blob 99134ec7f247 done
Copying blob c3a984abe8a8 done
Copying config 3f9bb7a003 done
Writing manifest to image destination
Storing signatures

# Try pulling encrypted image without key
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah pull --tls-verify=false docker://localhost:5000/test_enc_img
Getting image source signatures
Getting image source signatures
Getting image source signatures
Getting image source signatures
while pulling "docker://localhost:5000/test_enc_img" as "localhost:5000/test_enc_img": Error decrypting layer sha256:ae69626ca4b44e82dce805f5f7be5272c5651bd75e19aa38ec2f5ef78986f3ff: missing private key needed for decryption
ERRO exit status 1

# Try pulling encrypted image with key
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah pull --tls-verify=false --decryption-key ./mykey.pem docker://localhost:5000/test_enc_img
Getting image source signatures
Copying blob ae69626ca4b4 done
Copying blob de5dd0ed1059 done
Copying blob 4071791f0baf done
Copying config 3f9bb7a003 done
Writing manifest to image destination
Storing signatures
3f9bb7a003b412bc186a8437e2d2246f42a6bb7488458585948e50be00a9037d

# Cleanup
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah rmi --all -f
ed21b7a8aee9cc677df6d7f38a641fa0e3c05f65592c592c9f28c42b3dd89291
3f9bb7a003b412bc186a8437e2d2246f42a6bb7488458585948e50be00a9037d

# Create dockerfile that uses encrypted image
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ vi enc_dockerfile/Dockerfile
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ cat enc_dockerfile/Dockerfile
FROM localhost:5000/test_enc_img

# buildah bud without key
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah bud --tls-verify=false ./enc_dockerfile/
STEP 1: FROM localhost:5000/test_enc_img
Getting image source signatures
Getting image source signatures
Getting image source signatures
Getting image source signatures
error creating build container: Error decrypting layer sha256:ae69626ca4b44e82dce805f5f7be5272c5651bd75e19aa38ec2f5ef78986f3ff: missing private key needed for decryption
ERRO exit status 1

# buildah bud with key
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ ./buildah bud --tls-verify=false --decryption-key ./mykey.pem ./enc_dockerfile/
STEP 1: FROM localhost:5000/test_enc_img
Getting image source signatures
Copying blob ae69626ca4b4 done
Copying blob 4071791f0baf done
Copying blob de5dd0ed1059 done
Copying config 3f9bb7a003 done
Writing manifest to image destination
Storing signatures
STEP 2: COMMIT
--> 3f9bb7a003b
3f9bb7a003b412bc186a8437e2d2246f42a6bb7488458585948e50be00a9037d
```

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Uncomment the following comment block and include the issue
number or None on one line.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`, or `None`.
-->
None

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

Yes 
- it provides a new flag for `buildah push` with `--encryption-key`
- it provides a new flag for `buildah pull` with `--decryption-key`
- it provides a new flag for `buildah bud` with `--decryption-key`

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes please follow the kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note

```

lumjjb commented:
This is the initial implementation idea. There are a few more things that I am working on right now before taking out of DRAFT, but would like some feedback on the general idea.

TODO in progress by me:
- Remove ufave/cli and some additional imports which are not necessary from upstream ocicrypt lirbary 
- Add tests

lumjjb commented:
Thanks for comments @rhatdan ! Addressed them. 

>command completions changes required.

Forgive my ignorance.. Is there a process to do this? Or is this a manual edit of `contrib/completions/bash/buildah`?

rhatdan commented:
Manually edit them.  Just add the options.

TomSweeneyRedHat commented:
Concept looks good to me, nothing out of whack that I spotted.  We will most likely need to do this for at least 'podman pull' too, maybe others as well.  The build side of things should go in without a problem, but we'll need to touch up the `podman build` man page to capture the new options.

FYI: @mheon @baude

lumjjb commented:
i'm running into some weird flag issues in running the tests out of `master`. Having issues with running the integration tests. Am i missing some configuration? I get similar errors when I run `make test-integration`.


```bash
# bats ./tests/push.bat
...
 ✗ push
...
   $ /home/vagrant/go/src/github.com/containers/buildah/tests/../buildah commit
   -D
   --format docker
   --reference-time
   /tmp/tmpdbb3572b5a974dc1a0dd361f/reference-time-file
   --signature-policy
   /home/vagrant/go/src/github.com/containers/buildah/tests/policy.json
   working-container
   scratch-image-docker
   unknown flag: --format docker
   [ rc=1 (** EXPECTED 0 **) ]
   #/vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
   #| FAIL: exit code is 1; expected 0
   #\^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...

 ✗ buildah push image to docker and docker registry
   (from function `_prefetch' in file tests/helpers.bash, line 71,
    in test file tests/push.bats, line 148)
     `_prefetch busybox' failed with status 125
   # [checking for: busybox]
   # [restoring from cache: /tmp/buildah-image-cache.4027 / busybox]
   Error: unknown flag: --root /tmp/tmp9db9842473732348fdfd7845/root --storage-driver vfs

```

rhatdan commented:
No idea what is going on.

lumjjb commented:
looks like it was a mismatch version with podman , should be good now.

lumjjb commented:
@rhatdan figured it out... looks like it was an outdated `bats` version which was parsing the arguments in a different way.

lumjjb commented:
@rhatdan @TomSweeneyRedHat Added the tests and `buildah from`. Ready for review!

rhatdan commented on cmd/buildah/bud.go:
```diff
@@ -66,6 +68,7 @@ func init() {
 	// BUD is a all common flags
 	budFlags := buildahcli.GetBudFlags(&budFlagResults)
 	budFlags.StringVar(&budFlagResults.Runtime, "runtime", util.Runtime(), "`path` to an alternate runtime. Use BUILDAH_RUNTIME environment variable to override.")
+	flags.StringSliceVar(&budFlagResults.DecryptionKeys, "decryption-key", nil, "*Experimental* key needed to decrypt the image")
```
Use lower case  for help messages.. But I would drop "Experimental" and just leave that in the man page.

rhatdan commented on cmd/buildah/bud.go:
```diff
@@ -289,6 +292,17 @@ func budCmd(c *cobra.Command, inputArgs []string, iopts budOptions) error {
+		// decryption
+		dcc, err := enchelpers.CreateCryptoConfig([]string{}, iopts.BudResults.DecryptionKeys)
+		if err != nil {
+			return errors.Wrapf(err, "Invalid decryption keys")
```
Lower case error.

rhatdan commented on cmd/buildah/from.go:
```diff
@@ -72,6 +75,7 @@ func init() {
 	flags.BoolVar(&opts.pullAlways, "pull-always", false, "pull the image even if the named image is present in store")
 	flags.BoolVar(&opts.pullNever, "pull-never", false, "do not pull the image, use the image present in store if available")
 	flags.BoolVarP(&opts.quiet, "quiet", "q", false, "don't output progress information when pulling images")
+	flags.StringSliceVar(&opts.decryptionKeys, "decryption-key", nil, "*experimental* key needed to decrypt the image")
```
Remove experimental

rhatdan commented on cmd/buildah/from.go:
```diff
@@ -262,6 +266,18 @@ func fromCmd(c *cobra.Command, args []string, iopts fromReply) error {
+		// decryption
+		dcc, err := enchelpers.CreateCryptoConfig([]string{}, iopts.decryptionKeys)
+		if err != nil {
+			return errors.Wrapf(err, "Invalid decryption keys")
```
lower case error.

rhatdan commented on cmd/buildah/pull.go:
```diff
@@ -57,6 +60,7 @@ func init() {
 	flags.StringVar(&opts.creds, "creds", "", "use `[username[:password]]` for accessing the registry")
 	flags.BoolVarP(&opts.removeSignatures, "remove-signatures", "", false, "don't copy signatures when pulling image")
 	flags.StringVar(&opts.signaturePolicy, "signature-policy", "", "`pathname` of signature policy file (not usually used)")
+	flags.StringSliceVar(&opts.decryptionKeys, "decryption-key", nil, "*experimental* key needed to decrypt the image")
```
remove experimental

rhatdan commented on cmd/buildah/pull.go:
```diff
@@ -101,6 +105,17 @@ func pullCmd(c *cobra.Command, args []string, iopts pullOptions) error {
+		// decryption
+		dcc, err := enchelpers.CreateCryptoConfig([]string{}, iopts.decryptionKeys)
+		if err != nil {
+			return errors.Wrapf(err, "Invalid decryption keys")
```
Lowercase description.  Should this be made into a shared function, since we repeated it many times?

lumjjb replied:
Yea - agreed. let me refactor this.

rhatdan commented on cmd/buildah/push.go:
```diff
@@ -76,6 +80,9 @@ func init() {
 	flags.BoolVarP(&opts.removeSignatures, "remove-signatures", "", false, "don't copy signatures when pushing image")
 	flags.StringVar(&opts.signBy, "sign-by", "", "sign the image using a GPG key with the specified `FINGERPRINT`")
 	flags.StringVar(&opts.signaturePolicy, "signature-policy", "", "`pathname` of signature policy file (not usually used)")
+	flags.StringSliceVar(&opts.encryptionKeys, "encryption-key", nil, "*experimental* key with the encryption protocol to use needed to encrypt the image (e.g. jwe:/path/to/key.pem)")
```
remove experimental.

rhatdan commented on cmd/buildah/push.go:
```diff
@@ -76,6 +80,9 @@ func init() {
 	flags.StringVar(&opts.signBy, "sign-by", "", "sign the image using a GPG key with the specified `FINGERPRINT`")
 	flags.StringVar(&opts.signaturePolicy, "signature-policy", "", "`pathname` of signature policy file (not usually used)")
+	flags.StringSliceVar(&opts.encryptionKeys, "encryption-key", nil, "*experimental* key with the encryption protocol to use needed to encrypt the image (e.g. jwe:/path/to/key.pem)")
+	flags.IntSliceVar(&opts.encryptLayers, "encrypt-layer", nil, "*experimental* the 0-indexed layer indices, with support for negative indexing (e.g. 0 is the first layer, -1 is the last layer), if not defined, will encrypt all layers if encryption-key flag is specified")
```
Ditto

rhatdan commented on cmd/buildah/push.go:
```diff
@@ -164,6 +171,20 @@ func pushCmd(c *cobra.Command, args []string, iopts pushOptions) error {
+		encLayers = &iopts.encryptLayers
+		ecc, err := enchelpers.CreateCryptoConfig(iopts.encryptionKeys, []string{})
+		if err != nil {
+			return errors.Wrapf(err, "Invalid encryption keys")
```
lower case

rhatdan commented on commit.go:
```diff
@@ -132,6 +133,19 @@ type PushOptions struct {
 	// RetryDelay is how long to wait before retrying a push attempt.
 	RetryDelay time.Duration
+
+	// If OciEncryptConfig is non-nil, it indicates that an image should be encrypted.
```
This should start with OciEncryptConfig not If.

rhatdan commented on commit.go:
```diff
@@ -319,7 +333,8 @@ func (b *Builder) Commit(ctx context.Context, dest types.ImageReference, options
 	var manifestBytes []byte
-	if manifestBytes, err = retryCopyImage(ctx, policyContext, maybeCachedDest, maybeCachedSrc, dest, "push", getCopyOptions(b.store, options.ReportWriter, nil, systemContext, "", false, options.SignBy), options.MaxRetries, options.RetryDelay); err != nil {
+	logrus.Debugf("Calling from Commit")
+	if manifestBytes, err = retryCopyImage(ctx, policyContext, maybeCachedDest, maybeCachedSrc, dest, "push", getCopyOptions(b.store, options.ReportWriter, nil, systemContext, "", false, options.SignBy, nil, nil, nil), options.MaxRetries, options.RetryDelay); err != nil {
```
Rather then extending retryCopyImage, could you just pass in options...

lumjjb replied:
I think we're extending `getCopyOptions` here. Is this fine or are you thinking something like this?

```
options := getCopyOptions(...)
options.decryptionKey = ...
```

rhatdan replied:
I ready this wrong.  I was thinking the copy options and options.MaxRetries,  I hate functions which just keep growing their params.  Is there a way we can consolidate.

lumjjb replied:
Have a couple ideas ... But seems like this would be a separate PR.


For `getCopyOptions` it looks like the only special case is the `store`... Maybe something like this? I guess it would be a different PR for this though.

```
initCopyOptionsWithStore(store, &cp.Options{
                ReportWriter:          reportWriter,
                SourceCtx:             sourceCtx,
                DestinationCtx:        destinationCtx,
                ForceManifestMIMEType: manifestType,
                RemoveSignatures:      removeSignatures,
                SignBy:                addSigner,
                OciEncryptConfig:      ociEncryptConfig,
                OciDecryptConfig:      ociDecryptConfig,
                OciEncryptLayers:      ociEncryptLayers,
        })
```


Consolidating the retry options
```
type copyRetryOptions struct{
    maxRetries int
    retryDelay time.Duration
}
```

rhatdan commented:
@mtrmac @vrothberg @nalind PTAL

rhatdan commented:
getEncryptConfig is not used?
This looks good, but we will want you to squash the commits when you are done.

lumjjb commented:
Thanks. good catch. Rebased master and squashed as well.

mtrmac commented on commit.go:
```diff
@@ -319,7 +333,8 @@ func (b *Builder) Commit(ctx context.Context, dest types.ImageReference, options
 
 	var manifestBytes []byte
-	if manifestBytes, err = retryCopyImage(ctx, policyContext, maybeCachedDest, maybeCachedSrc, dest, "push", getCopyOptions(b.store, options.ReportWriter, nil, systemContext, "", false, options.SignBy), options.MaxRetries, options.RetryDelay); err != nil {
+	logrus.Debugf("Calling from Commit")
```
Is this intended to remain?

lumjjb replied:
Nope. let me remove this. tks

mtrmac commented on commit.go:
```diff
@@ -319,7 +333,8 @@ func (b *Builder) Commit(ctx context.Context, dest types.ImageReference, options
 	var manifestBytes []byte
-	if manifestBytes, err = retryCopyImage(ctx, policyContext, maybeCachedDest, maybeCachedSrc, dest, "push", getCopyOptions(b.store, options.ReportWriter, nil, systemContext, "", false, options.SignBy), options.MaxRetries, options.RetryDelay); err != nil {
+	logrus.Debugf("Calling from Commit")
+	if manifestBytes, err = retryCopyImage(ctx, policyContext, maybeCachedDest, maybeCachedSrc, dest, "push", getCopyOptions(b.store, options.ReportWriter, nil, systemContext, "", false, options.SignBy, nil, nil, nil), options.MaxRetries, options.RetryDelay); err != nil {
```
It is possible to commit directly to a remote registry, so the commit command should probably contain the same encryption options as push.

lumjjb replied:
My impression was that commit was to create an image locally and then push to export it to a registry or files. My thought was that locally committed images would be decrypted and unpacked so that they can be run via the CLI, and so any images that are allowed to be used locally to build on top of should remain decrypted. I'm not too sure about specifics about commit.

mtrmac replied:
I agree it’s not the most common usage, but it is explicitly documented e.g. at https://github.com/containers/buildah/blob/master/docs/buildah-commit.md#example .

lumjjb replied:
Let me dig a bit deeper. There may be some weird interactions, especially with buildah run... Unless we want to special case on registry for commit?

mtrmac replied:
There is already an `options.SignBy` on this line, so I imagine the encryption options could get from the CLI to this place using a similar route. But I’m only minimally familiar with the Buildah codebase.

lumjjb replied:
Okay i think i misunderstood how the running of contianers works here.

I think it looks good, the only issue i seem to be running into an issue with `blobcache`. Works fine otherwise 

local image
```
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df test
Getting image source signatures
Copying blob 5f70bf18a086 done
error committing container "contid" to "test": error copying layers and metadata for container "b606f78574df9ddfefc04d0a98f76fa2ae97a1ebaed1df5cde375576c692be8c": Error reading blob sha256:5b0d2d635df829f65d0ffb45eab2c3124a470c4f385d6602bda0c21c5248bcab: error opening file "/var/tmp/buildah112495834/sha256:5b0d2d635df829f65d0ffb45eab2c3124a470c4f385d6602bda0c21c5248bcab" to buffer layer blob: open /var/tmp/buildah112495834/sha256:5b0d2d635df829f65d0ffb45eab2c3124a470c4f385d6602bda0c21c5248bcab: no such file or directory
```

Other cases
```
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df docker://localhost:5000/test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config b0924b4f50 done
Writing manifest to image destination
Storing signatures
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df oci:test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config 8959f8f5d5 done
Writing manifest to image destination
Storing signatures
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df dir:test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config 5ded513e65 done
Writing manifest to image destination
Storing signatures

```

lumjjb replied:
Ok i fixed the above error... but when I commit to a local image. it is trying to unpack the image afterwards. Let me see if i can commit it without unpacking.

```
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah commit --tls-verify=false --encryption-key jwe:/home/vagrant/mypubkey.pem  b606f78574df test
Getting image source signatures
Copying blob 5b0d2d635df8 done
Copying blob 5f70bf18a086 done
Copying config 45c0e7b514 done
Writing manifest to image destination
Storing signatures
error committing container "contid" to "test": error copying layers and metadata for container "b606f78574df9ddfefc04d0a98f76fa2ae97a1ebaed1df5cde375576c692be8c": Error committing the finished image: error adding layer with blob "sha256:c41eabc97b5b0885d97663759516f07af539ba887b5d78a4e12e26b681aca7d7": Error processing tar file(exit status 1): archive/tar: invalid tar header
```

lumjjb replied:
@mtrmac  It looks like (but please correct me) that the decision to unpack is being done based on the `ImageDestination`. For this specific case, I believe it is from this function which is done during `ImageDestination.Commit` via `PutLayer`.

https://github.com/containers/image/blob/master/storage/storage_image.go#L743

It looks possible to add a check around here since the manifest is available, we can skip the unpacking of encrypted layers by checking the mediatype:

```
if strings.HasSuffix(layer.BlobInfo.Mediatype, "+encrypted") {
    continue
}
```
https://github.com/containers/image/blob/master/storage/storage_image.go#L652

The issue with this would be that not all container images that are in this particular store would be runnable. Would this affect the expected semantics of this `ImageDestination`.

Alternatively, in the meantime I can do `buildah commit` that will implement encryption for everything except local images, in which case it will throw an error on the `buildah` level

mtrmac replied:
> @mtrmac It looks like (but please correct me) that the decision to unpack is being done based on the `ImageDestination`.

By “unpack”, do you mean “decompress” or “extract an archive to individual files”?

> For this specific case, I believe it is from this function which is done during `ImageDestination.Commit` via `PutLayer`.
> 
> https://github.com/containers/image/blob/master/storage/storage_image.go#L743
> 
> It looks possible to add a check around here since the manifest is available, we can skip the unpacking of encrypted layers by checking the mediatype:
> 
> ```
> if strings.HasSuffix(layer.BlobInfo.Mediatype, "+encrypted") {
>     continue
> }
> ```

That makes no sense to me. c/storage only represents layers as filesystem trees, and images as a series of layers on top of each other. If some “encrypted” layer is not recorded as a filesystem tree, it plainly does not exist and therefore an image containing that layer can not exist in c/storage either.

It does seem that we forgot to teach c/image that some transports can only work with unencrypted layers, and refuse to copy encrypted layers without decrypting; but that should happen in the generic code, the same way transports declare `AcceptsForeignURLs` they should declare they (in?)ability to accept encrypted layers.

Either way, this does not quite help Buildah…

---

> Alternatively, in the meantime I can do `buildah commit` that will implement encryption for everything except local images, in which case it will throw an error on the `buildah` level

Yes, I think something like that should happen. It is currently, AFAIK, plainly impossible to represent an encrypted image when committing to local storage, so Buildah should not be trying. (It may well be good enough to just add the check in c/image and have Buildah report the error, but reporting a better-targeted error when the user asks for a nonsensical operation could be more helpful.)

---

(All of this being said, it’s been quite some time since I’ve last read the blob cache code, and I’m not quite sure it can work with encryption. But it seems you are testing all of that already.)

mtrmac commented on docs/buildah-bud.md:
```diff
@@ -178,6 +178,10 @@ The [username[:password]] to use to authenticate with the registry if required.
 
+**--decryption-key** *key*
+
+A reference required to perform decryption of container images. This should point to files which represent keys and/or certificates that can be used for decryption. Decryption will be tried with all keys. This feature is still *experimental*.
```
“reference”?

lumjjb replied:
I think in the context of that it's because its pointing to a file and not the key data itself. This is made to be consistent with the skopeo docs.

mtrmac replied:
Huh, so it is consistent with Skopeo. I haven’t noticed :)

Basically I think “A reference required to perform decryption of container images. This should point to files which represent… that can be used for decryption” is just a very long way to say “Path to a file containing… required for decryption” or something like that. 

I don’t see any additional value gained by adding “reference”, “point to”, “represent”.

lumjjb replied:
Yea.. i think the way this got constructed was because the way encryption-key is specified is that it requires `<protocol>:<file>`... But I do agree, would be a good opportunity to reword this. I can update this and then once this gets merged ill update skopeo docs.

mtrmac commented on docs/buildah-pull.md:
```diff
@@ -62,6 +62,10 @@ The [username[:password]] to use to authenticate with the registry if required.
 
+**--decryption-key** *key*
+
+A reference required to perform decryption of container images. This should point to files which represent keys and/or certificates that can be used for decryption. Decryption will be tried with all keys. This feature is still *experimental*.
```
“reference”?

mtrmac commented on docs/buildah-push.md:
```diff
@@ -66,6 +66,10 @@ After copying the image, write the digest of the resulting image to the file.
 
+**--encryption-key** *key*
+
+A reference prefixed with the encryption protocol to use. The supported protocols are JWE, PGP and PKCS7. For instance, jwe:/path/to/key.pem or pgp:admin@example.com or pkcs7:/path/to/x509-file. This feature is still *experimental*.
```
“reference”?

TomSweeneyRedHat replied:
It would be nice to have an explanation of JWE, PGP and PKCS7 and/or man page references.  Ditto through out.

lumjjb replied:
Added RFC reference numbers for the protocols. Would you like to add the reference to https://github.com/opencontainers/artifacts/pull/15 for context? Or will the RFCs be sufficient for this?

**--encryption-key** *key*

The [protocol:keyfile] to use for image encryption. The protocol can specifies the encryption protocol, which can be  JWE(RFC7516), PGP(RFC4880), and PKCS7(RFC2315). For instance, jwe:/path/to/key.pem or pgp:admin@example.com or pkcs7:/path/to/x509-file.

lumjjb commented:
Looks like this may potentially be an issue with ubuntu 19.04 test:
```${SCRIPT_BASE}/setup.sh |& ${_TIMESTAMP}
[16:09:12] START - All [+xxxx] lines that follow are relative to right now.
[+0000s] Installing script to mask stdout/stderr unless non-zero exit.
[+0000s] Setting up ubuntu 19
[+0000s] Retrying 5 times with a 30s delay, and 24s timeout for command: apt-get -qq --yes update
[+0158s] ##### (attempt #1)
[+0158s] E: The repository 'http://us-central1.gce.archive.ubuntu.com/ubuntu disco Release' no longer has a Release file.
[+0158s] E: The repository 'http://us-central1.gce.archive.ubuntu.com/ubuntu disco-updates Release' no longer has a Release file.
[+0158s] E: The repository 'http://us-central1.gce.archive.ubuntu.com/ubuntu disco-backports Release' does not have a Release file.
[+0158s] E: The repository 'http://security.ubuntu.com/ubuntu disco-security Release' no longer has a Release file.
[+0158s] ##### (failed with exit: 100)```

rhatdan commented:
We are seeing the Ubuntu issue everywhere in github.com/containers.

lumjjb commented:
/retest

lumjjb commented:
Tests passed!

rhatdan commented:
@mtrmac PTAL

lumjjb commented:
Rebased on master

rhatdan commented:
@nalind @mtrmac @vrothberg @TomSweeneyRedHat PTAL

TomSweeneyRedHat commented on tests/pull.bats:
```diff
@@ -167,3 +167,60 @@ load helpers
+  run_buildah rmi localhost:5000/buildah/busybox_encrypted:latest
+
+  rm -rf ${TESTDIR}/tmp
+}
```
I tried building a public/private key with a passphrase.  The push went ok, but when I pulled, I got the below error.  Is there a way to include the passphrase for the private key?

```
# buildah push --encryption-key jwe:public.pem alpine docker://localhost:5000/alpine_encrypted1:latest

# buildah pull --decryption-key private.pem docker://localhost:5000/alpine_encrypted:latest
unable to obtain decrypt config: invalid decryption keys: : Missing password for encrypted private key
```

mtrmac replied:
Apparently this is implemented using the Go openpgp package.

And there seems to be a syntax for passing passphrases, looking at https://github.com/containers/ocicrypt/blob/bdddf7ce0bed9bead019e491cc7e781d5c7d8c4b/helpers/parse_helpers.go#L100-L107 . That definitely should have user-facing documentation _somewhere_, and the man pages should point to it.

lumjjb replied:
Yup - I will update the documentation. Just tested locally:

```
# Without passphrase
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah --tls-verify=false pull --decryption-key privkey.pem docker://localhost:5000/alpine_encrypted1:latest
unable to obtain decrypt config: invalid decryption keys: : Missing password for encrypted private key

# With passphrase
vagrant@ubuntu-bionic:~/go/src/github.com/containers/buildah$ sudo ./buildah --tls-verify=false pull --decryption-key privkey.pem:foobar docker://localhost:5000/alpine_encrypted1:latest
Getting image source signatures
Copying blob 3c83cf2f9e96 done
Copying config af88fdb253 done
Writing manifest to image destination
Storing signatures
af88fdb253aac46693de7883c9c55244327908c77248d7654841503f744aae8b```

lumjjb replied:
Updated the docs to: 


**--decryption-key** *key*

The [keyfile[:passphrase]] to be used for decryption of images. Keyfiles can point to keys and/or certificates.       Decryption will be tried with all keys. If the key is protected by a passphrase is required to be paseed in, and      should be ommitted otherwise.

TomSweeneyRedHat commented:
I too would like the "reference" worked out in the man pages per @mtrmac question.  I realize it's lifted from skopeo's.  Ditto a bit more clarity on what the keys are that are accepted.

I ran this through the baseline tests and that all passed without a problem.  I also played with the new functionality myself and had issues with private keys with a passphrase as noted in another comment.

I definitely would like a review from @nalind and/or @mtrmac if possible, but I'd be OK with this going in sooner rather than later given the "experimental" tags and with the exception of the passphrase, which might be a user error on my end, works well.

@lumjjb really nice PR overall, ty.

mtrmac commented on commit.go:
```diff
@@ -88,6 +89,17 @@ type CommitOptions struct {
 	RetryDelay time.Duration
+	// OciEncryptConfig when non-nil indicates that an image should be encrypted.
+	// The encryption options is derived from the construction of EncryptConfig object.
+	// Note: During initial encryption process of a layer, the resultant digest is not known
+	// during creation, so newDigestingReader has to be set with validateDigest = false
```
This copy&pasted comment does not apply to this codebase.

mtrmac commented on commit.go:
```diff
@@ -132,6 +144,19 @@ type PushOptions struct {
+
+	// OciEncryptConfig when non-nil indicates that an image should be encrypted.
+	// The encryption options is derived from the construction of EncryptConfig object.
+	// Note: During initial encryption process of a layer, the resultant digest is not known
+	// during creation, so newDigestingReader has to be set with validateDigest = false
```
This comment does not apply.

mtrmac commented on commit.go:
```diff
@@ -270,7 +295,9 @@ func (b *Builder) Commit(ctx context.Context, dest types.ImageReference, options
 	exportBaseLayers := true
-	if transport, destIsStorage := dest.Transport().(is.StoreTransport); destIsStorage && b.FromImageID != "" {
+	if transport, destIsStorage := dest.Transport().(is.StoreTransport); destIsStorage && options.OciEncryptConfig != nil {
+		return imgID, nil, "", errors.New("unable to use local storage with image encryption")
```
(This should end up as a native functionality in c/image eventually, somehow dealing with API stability. But this check can give users better error messages, so it’s fine to leave it in.)

lumjjb commented:
Thanks @TomSweeneyRedHat @mtrmac for the comments, all valid points. I will incorporate the feedback - and keep track of the changes and create apply those changes to skopeo as well after this PR.

lumjjb commented:
@mtrmac @TomSweeneyRedHat Addressed the comments! 

I am thinking that I could do another PR after this one to write something in `docs/tutorials` on image security - including both signing and encryption as part of that tutorial to provide more context.

TomSweeneyRedHat commented on docs/buildah-push.md:
```diff
@@ -66,6 +66,10 @@ After copying the image, write the digest of the resulting image to the file.
+**--encryption-key** *key*
+
+The [protocol:keyfile] specifies the encryption protocol, which can be JWE (RFC7516), PGP (RFC4880), and PKCS7 (RFC2315) and the key material required for image encryption. For instance, jwe:/path/to/key.pem or pgp:admin@example.com or pkcs7:/path/to/x509-file.
+
```
:+1:

TomSweeneyRedHat commented on docs/buildah-bud.md:
```diff
@@ -178,6 +178,10 @@ The [username[:password]] to use to authenticate with the registry if required.
 If one or both values are not supplied, a command line prompt will appear and the
 value can be entered.  The password is entered without echo.
 
+**--decryption-key** *key*
```
I'd suggest the following, ditto for the other man pages with decryption-key
```suggestion
**--decryption-key** *key[:passphrase]*
```

TomSweeneyRedHat commented:
One nit scattered across several of the pages, otherwise LGTM.
A follow up tutorial and/or blog for buildah.io would be very welcomed!

lumjjb commented:
Fixed!

rhatdan commented:
Tests are still failing?

lumjjb commented:
Interesting, it was just passing the last commit and just doc changes... let me check it out.

lumjjb commented:
Ah - seems to have something to do with Quay.io being down.

`[+0028s] Error: unable to pull quay.io/libpod/in_podman:master: unable to pull image: Error initializing source docker://quay.io/libpod/in_podman:master: unexpected http code: 500 (Internal Server Error), URL: https://quay.io/v2/auth?scope=repository%3Alibpod%2Fin_podman%3Apull&service=quay.io
`

TomSweeneyRedHat commented:
bors retest

TomSweeneyRedHat commented:
Kicked the test runs again, hopefully quay.io is happy again.

rhatdan commented:
bors r+

lumjjb commented:
🎉

TomSweeneyRedHat commented:
Awesome work @lumjjb !
